staging/vhost-device-gpu/src/main.rs

```diff
@@ -0,0 +1,113 @@
+// VIRTIO GPU Emulation via vhost-user
+//
+// Copyright 2024 Red Hat Inc
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use log::{error, info};
+use std::{path::PathBuf, process::exit};
+
+use clap::Parser;
+use thiserror::Error as ThisError;
+use vhost_device_gpu::{
+    device::{self, VhostUserGpuBackend},
+    GpuConfig,
+};
+use vhost_user_backend::VhostUserDaemon;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+pub(crate) enum Error {
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(device::Error),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct GpuArgs {
+    /// vhost-user Unix domain socket.
+    #[clap(short, long, value_name = "SOCKET")]
+    socket_path: PathBuf,
+}
+
+impl TryFrom<GpuArgs> for GpuConfig {
+    type Error = Error;
```

Comment:
Ok, that's reasonable. If you're planning on adding logic that can make the conversion return an `Err` then `Infallible` is not applicable here indeed.

Reply:
This never fails still so it should switch to using Infalliable or just use From trait instead of TryFrom.

Reply:
well, it can fail when you try provide a wrong path, see tests below.

Reply:
It fails when calling start_backend, not here.

Reply:
this directly affects the error enum for handling start_backend, changing the error type to infallible https://github.com/rust-vmm/vhost-device/blob/220b091da1a12874a72996baa7807d61f3641115/staging/vhost-device-gpu/src/main.rs#L42 means output type has to match the trait: `std::result::Result<GpuConfig, Infallible>` https://github.com/rust-vmm/vhost-device/blob/220b091da1a12874a72996baa7807d61f3641115/staging/vhost-device-gpu/src/main.rs#L19 which directly implies that when `daemon.serve(socket).map_err(Error::ServeFailed)?;` returns an error we cannot use any of the error enum. that's because using infallible `pub enum Infallible {}` is `The error type for errors that can never happen.`

I can use the from trait

Reply:
You already call unwrap here https://github.com/rust-vmm/vhost-device/blob/220b091da1a12874a72996baa7807d61f3641115/staging/vhost-device-gpu/src/main.rs#L79C35-L79C80 anyways, so you can just use the From trait or maybe event construct `GpuConfig::new(socket_path, gpu_mode) ` directly there without implementing a trait that is only used once